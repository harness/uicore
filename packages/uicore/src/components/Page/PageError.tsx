/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { Text, Icon, Layout, Button, ButtonVariation, ButtonProps, Container } from '../../'
import { Color } from '@harness/design-system'
import React from 'react'
import i18n from './PageError.i18n'

export interface PageErrorProps {
  message?: React.ReactNode
  width?: number
  className?: string
  onClick?: ButtonProps['onClick']
  disabled?: boolean
}

const getErrorNode = (message: React.ReactNode): React.ReactNode => {
  if (!message) {
    return (
      <Text font={{ align: 'center' }} color={Color.RED_500}>
        {i18n.generalError}
      </Text>
    )
  }
  if (typeof message === 'string') {
    return (
      <Text font={{ align: 'center' }} color={Color.RED_500}>
        {message}
      </Text>
    )
  }
  return message
}

export const PageError: React.FC<PageErrorProps> = props => {
  return (
    <Container width="100%" height="100%" flex={{ align: 'center-center' }}>
      <Layout.Vertical
        spacing="medium"
        width={props?.width || 500}
        style={{ alignItems: 'center' }}
        className={props.className}>
        <Icon name="error" size={32} color={Color.RED_500} />
        {getErrorNode(props.message)}
        {props.onClick && (
          <Button
            variation={ButtonVariation.PRIMARY}
            text={i18n.retry}
            onClick={props.onClick}
            disabled={props.disabled}
          />
        )}
      </Layout.Vertical>
    </Container>
  )
}
