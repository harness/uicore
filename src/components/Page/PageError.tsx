/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { Color, Text, Icon, Layout, Button, ButtonVariation, ButtonProps, Container } from '../../'
import React from 'react'
import i18n from './PageError.i18n'

export interface PageErrorProps {
  message?: string
  width?: number
  className?: string
  onClick?: ButtonProps['onClick']
  disabled?: boolean
}

export const PageError: React.FC<PageErrorProps> = props => (
  <Container width="100%" height="100%" flex={{ align: 'center-center' }}>
    <Layout.Vertical
      spacing="medium"
      width={props?.width || 500}
      style={{ alignItems: 'center' }}
      className={props.className}>
      <Icon name="error" size={32} color={Color.RED_500} />
      <Text font={{ align: 'center' }} color={Color.RED_500}>
        {props.message || i18n.generalError}
      </Text>
      {props.onClick && (
        <Button
          variation={ButtonVariation.PRIMARY}
          text={i18n.retry}
          onClick={props.onClick}
          disabled={props.disabled}
        />
      )}
    </Layout.Vertical>
  </Container>
)
