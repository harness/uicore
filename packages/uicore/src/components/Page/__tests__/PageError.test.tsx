/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { Layout, Text } from 'index'
import { render } from '@testing-library/react'
import { PageError } from '../PageError'

describe('PageError test', () => {
  test('Rendering error with no message', () => {
    const { container } = render(<PageError />)
    expect(container).toMatchSnapshot()
  })

  test('Rendering error with a string', () => {
    const { container } = render(<PageError message="Invalid Request" />)
    expect(container).toMatchSnapshot()
  })

  test('Rendering error with a React Node', () => {
    const reactNode = (): React.ReactElement => (
      <Layout.Vertical>
        <Text>You are missing the following permission:</Text>
      </Layout.Vertical>
    )
    const { container } = render(<PageError message={reactNode()} />)
    expect(container).toMatchSnapshot()
  })
})
