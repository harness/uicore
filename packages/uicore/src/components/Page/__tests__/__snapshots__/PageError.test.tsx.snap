// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageError test Rendering error with a React Node 1`] = `
<div>
  <div
    class="StyledProps--font StyledProps--main StyledProps--flex StyledProps--flex-align-center-center"
    style="width: 100%; height: 100%;"
  >
    <div
      class="vertical layout-spacing-medium StyledProps--main"
      style="align-items: center; width: 500px;"
    >
      <span
        class="bp3-icon bp3-icon-error StyledProps--main StyledProps--color StyledProps--color-red500"
        icon="error"
      >
        <svg
          data-icon="error"
          height="32"
          viewBox="0 0 20 20"
          width="32"
        >
          <desc>
            error
          </desc>
          <path
            d="M10 0C4.48 0 0 4.48 0 10s4.48 10 10 10 10-4.48 10-10S15.52 0 10 0zm1 16H9v-2h2v2zm0-3H9V4h2v9z"
            fill-rule="evenodd"
          />
        </svg>
      </span>
      <div
        class="vertical layout-spacing-undefined StyledProps--main"
      >
        <p
          class="StyledProps--font StyledProps--main"
        >
          You are missing the following permission:
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`PageError test Rendering error with a string 1`] = `
<div>
  <div
    class="StyledProps--font StyledProps--main StyledProps--flex StyledProps--flex-align-center-center"
    style="width: 100%; height: 100%;"
  >
    <div
      class="vertical layout-spacing-medium StyledProps--main"
      style="align-items: center; width: 500px;"
    >
      <span
        class="bp3-icon bp3-icon-error StyledProps--main StyledProps--color StyledProps--color-red500"
        icon="error"
      >
        <svg
          data-icon="error"
          height="32"
          viewBox="0 0 20 20"
          width="32"
        >
          <desc>
            error
          </desc>
          <path
            d="M10 0C4.48 0 0 4.48 0 10s4.48 10 10 10 10-4.48 10-10S15.52 0 10 0zm1 16H9v-2h2v2zm0-3H9V4h2v9z"
            fill-rule="evenodd"
          />
        </svg>
      </span>
      <p
        class="StyledProps--font StyledProps--main StyledProps--font-align-center StyledProps--color StyledProps--color-red500"
      >
        Invalid Request
      </p>
    </div>
  </div>
</div>
`;

exports[`PageError test Rendering error with no message 1`] = `
<div>
  <div
    class="StyledProps--font StyledProps--main StyledProps--flex StyledProps--flex-align-center-center"
    style="width: 100%; height: 100%;"
  >
    <div
      class="vertical layout-spacing-medium StyledProps--main"
      style="align-items: center; width: 500px;"
    >
      <span
        class="bp3-icon bp3-icon-error StyledProps--main StyledProps--color StyledProps--color-red500"
        icon="error"
      >
        <svg
          data-icon="error"
          height="32"
          viewBox="0 0 20 20"
          width="32"
        >
          <desc>
            error
          </desc>
          <path
            d="M10 0C4.48 0 0 4.48 0 10s4.48 10 10 10 10-4.48 10-10S15.52 0 10 0zm1 16H9v-2h2v2zm0-3H9V4h2v9z"
            fill-rule="evenodd"
          />
        </svg>
      </span>
      <p
        class="StyledProps--font StyledProps--main StyledProps--font-align-center StyledProps--color StyledProps--color-red500"
      >
        We cannot perform your request at the moment. Please try again.
      </p>
    </div>
  </div>
</div>
`;
